File: FingoDriver/src/main/java/com/yelloco/fingodriver/utils/Storage.java
package com.yelloco.fingodriver.utils;

import android.content.Context;
import android.content.SharedPreferences;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.util.List;

public class Storage
{

    private static final Storage prefUtils = new Storage();
    private static SharedPreferences sharedPreferences;

    public static Storage getInstance() {
        return prefUtils;
    }

    public void initialize(Context context) {
        if (sharedPreferences == null) {
            sharedPreferences = context.getSharedPreferences("APP_PREF", Context.MODE_PRIVATE);
        }
    }

    public void storeString(String key, String value) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putString(key, value);
        editor.apply();
    }

    public void storeInt(String key, int value) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putInt(key, value);
        editor.apply();
    }

    public void storeLong(String key, long value) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putLong(key, value);
        editor.apply();

    }

    public void storeBoolean(String key, boolean value) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.putBoolean(key, value);
        editor.apply();
    }

    public String getString(String key) {
        return sharedPreferences.getString(key, null);
    }

    public long getLong(String key, long defaultValue)
    {
        return sharedPreferences.getLong(key, defaultValue);
    }

    public String getString(String key, String defValue) {
        return sharedPreferences.getString(key, defValue);
    }

    public int getInt(String key, int defValue) {
        return sharedPreferences.getInt(key, defValue);
    }

    public boolean getBoolean(String key, boolean defaultValue) {
        return sharedPreferences.getBoolean(key, defaultValue);
    }

    public boolean contains(String key) {
        return sharedPreferences.contains(key);
    }

    public void removeValue(String key) {
        SharedPreferences.Editor editor = sharedPreferences.edit();
        editor.remove(key);
        editor.apply();
    }

    public <T> void storeList(String key, List<T> list){
        Gson gson = new Gson();
        String s = gson.toJson(list);
        storeString(key, s);
    }

    public <T> List<T> getList(String key, TypeToken<List<T>> tt){
        Gson gson = new Gson();
        String string = getString(key);
        return gson.fromJson(string, tt.getType());
    }

    public void registerChangeListener(SharedPreferences.OnSharedPreferenceChangeListener onSharedPreferenceChangeListener)
    {
        sharedPreferences.registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener);
    }

    public void unregisterChangeListener(SharedPreferences.OnSharedPreferenceChangeListener onSharedPreferenceChangeListener) {
        sharedPreferences.unregisterOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener);
    }

    public void destroy(){
        sharedPreferences = null;
    }

}

File: FingoDriver/src/main/java/com/yelloco/fingodriver/utils/Storage.kt
package com.yelloco.fingodriver.utils

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.util.Log
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken

object Storage
{
    private const val TAG = "Storage"
    private var sharedPreferences: SharedPreferences? = null

    fun initialize(context: Context) {
        if (sharedPreferences == null) {
            sharedPreferences = context.getSharedPreferences("APP_PREF", Context.MODE_PRIVATE)
        }
    }

    fun storeString(key: String?, value: String?) {
        sharedPreferences?.let {
            val editor = it.edit()
            editor.putString(key, value)
            editor.apply()
        } ?: kotlin.run {
            Log.e(TAG, "Failed to storeString PREFS is NULL")
        }
    }

    fun storeInt(key: String?, value: Int) {
        sharedPreferences?.let {
            val editor = it.edit()
            editor.putInt(key, value)
            editor.apply()
        } ?: kotlin.run {
            Log.e(TAG, "Failed to storeInt PREFS is NULL")
        }
    }

    fun storeLong(key: String?, value: Long) {
        sharedPreferences?.let {
            val editor = it.edit()
            editor.putLong(key, value)
            editor.apply()
        } ?: kotlin.run {
            Log.e(TAG, "Failed to storeLong PREFS is NULL")
        }
    }

    fun storeBoolean(key: String?, value: Boolean) {
        sharedPreferences?.let {
            val editor = it.edit()
            editor.putBoolean(key, value)
            editor.apply()
        } ?: kotlin.run {
            Log.e(TAG, "Failed to storeBoolean PREFS is NULL")
        }
    }

    fun getString(key: String?): String? {
        return sharedPreferences?.getString(key, null) ?: kotlin.run {
            return null
        }
    }

    fun getLong(key: String?, defaultValue: Long): Long {
        return sharedPreferences?.getLong(key, defaultValue) ?: kotlin.run {
            return defaultValue
        }
    }

    fun getString(key: String?, defaultValue: String?): String? {
        return sharedPreferences?.getString(key, defaultValue) ?: kotlin.run {
            return defaultValue
        }
    }

    fun getInt(key: String?, defaultValue: Int): Int {
        return sharedPreferences?.getInt(key, defaultValue) ?: kotlin.run {
            Log.e(TAG, "Failed to getBoolean PREFS is NULL")
            return defaultValue
        }
    }

    fun getBoolean(key: String?, defaultValue: Boolean): Boolean {
        return sharedPreferences?.getBoolean(key, defaultValue) ?: kotlin.run {
            Log.e(TAG, "Failed to getBoolean PREFS is NULL")
            return defaultValue
        }
    }

    operator fun contains(key: String): Boolean {
        return sharedPreferences?.contains(key) ?: kotlin.run {
            return false
        }
    }

    fun removeValue(key: String?) {
        sharedPreferences?.let {
            val editor = it.edit()
            editor.remove(key)
            editor.apply()
        } ?: kotlin.run {
            Log.e(TAG, "Failed to removeValue PREFS is NULL")
        }
    }

    fun <T> storeList(key: String?, list: List<T>?) {
        val gson = Gson()
        val s = gson.toJson(list)
        storeString(key, s)
    }

    fun <T> getList(key: String?, tt: TypeToken<List<T>?>): List<T> {
        val gson = Gson()
        val string = getString(key)
        return gson.fromJson(string, tt.type)
    }

    fun registerChangeListener(onSharedPreferenceChangeListener: OnSharedPreferenceChangeListener?) {
        sharedPreferences?.registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener)
            ?: kotlin.run {
                Log.e(TAG, "Failed to registerChangeListener PREFS is NULL")
            }
    }

    fun unregisterChangeListener(onSharedPreferenceChangeListener: OnSharedPreferenceChangeListener?) {
        sharedPreferences?.unregisterOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener)
            ?: kotlin.run {
            Log.e(TAG, "Failed to unregisterChangeListener PREFS is NULL")
        }
    }

    fun destroy() {
        sharedPreferences = null
    }
}